File: 2023/cpp/41.cpp
#include <bits/stdc++.h>
using namespace std;

vector<string> winningNum, myNum;

void splitfunc(string card) 
{
    string part;
    vector<string> cp;
    istringstream iss(card);
    while (iss >> part) cp.emplace_back(part);

    vector<string>::iterator it;
    for (it = cp.begin()+2; *it != "|"; it++) {
        winningNum.emplace_back(*(it));
    }

    while (it != cp.end()) {
        myNum.emplace_back(*(it+1));
        it++;
    }
}

int main(void)
{
    ifstream inpfile;
    inpfile.open("../test-cases/inp4.txt");

    string card;
    vector<string> cards;
    while (getline(inpfile, card)) cards.emplace_back(card);

    int totalPoints = 0;

    for (auto card : cards) {
        int cardVal = 0;
        splitfunc(card);
        
        for (auto mn : myNum) {
            for (auto wn : winningNum) {
                if (wn == mn) {
                    cardVal = cardVal==0 ? 1 : cardVal*2;
                }
            }
        }

        winningNum.clear();
        myNum.clear();

        totalPoints += cardVal;
    }

    cout << "Total points are: " << totalPoints << endl;
    inpfile.close();
    return 0;
}

File: 2023/python/41.py
with open("../test-cases/inp4.txt", "r") as inpfile:
    cards = inpfile.readlines()
    totalPoints = 0
    
    for card in cards:
        cardVal = 0
        card = card.split(":")
        numbers = card[1].split("|")
        
        winning_numbers = [wnum for wnum in numbers[0].split() if wnum]
        my_numbers = [mnum for mnum in numbers[1].split() if mnum]
        
        for i in range(len(my_numbers)):
            for j in range(len(winning_numbers)):
                if winning_numbers[j] == my_numbers[i]:
                    cardVal = 1 if cardVal == 0 else cardVal*2
        
        totalPoints += cardVal
    
print(f"Total points are: {totalPoints}")
inpfile.close()